oddt/interactions.py

```diff
@@ -101,21 +98,20 @@ def hbond_acceptor_donor(mol1, mol2, cutoff=3.5, base_angle=120, tolerance=30):
         angle1 = angle(d['coords'][:, np.newaxis, :],
                        a['coords'][:, np.newaxis, :],
                        a['neighbors'])
-        a_neighbors_num = np.sum(~np.isnan(a['neighbors'][:, :, 0]), axis=-1)[:, np.newaxis]
         angle2 = angle(a['coords'][:, np.newaxis, :],
                        d['coords'][:, np.newaxis, :],
                        d['neighbors'])
-        d_neighbors_num = np.sum(~np.isnan(d['neighbors'][:, :, 0]), axis=-1)[:, np.newaxis]
-        strict = (((np.nan_to_num(angle1) > (base_angle / a_neighbors_num - tolerance)) |
-                   np.isnan(angle1)) &
-                  ((np.nan_to_num(angle2) > (base_angle / d_neighbors_num - tolerance)) |
-                   np.isnan(angle2))).all(axis=-1)
+        strict = ((np.nan_to_num(angle1) > (np.take(BASE_ANGLES, a['hybridization']) - tolerance)[:, np.newaxis]) &
```

Comment:
I think behaviour for NaNs changed and I'm not sure if it was intended. Previously we had `<condition> | np.isnan(angle)`, so it was always satisfied for NaN angle. Now we only have a serious of conditions, that won't be satisfied for NaNs (`np.nan_to_num(np.nan)` gives 0, so `>` will be false)

Reply:
Another thing is that's it's a very complex piece of code so it's easy to get lost. I think we should break it into smaller pieces.

First, I'd introduce a private helper function for checking a single array of angles, and break each check into steps inside this function:
```python
def _check_angles(angles, hybridizations, tolerance):
    """Helper function for checking if interactions are sctrict"""
    angles = np.nan_to_num(angles)   # if we really want to do it (see previous comment)

    ideal_angles = np.take(BASE_ANGLES, hybridizations)[:, np.newaxis]
    lower_bound = ideal_angles - tolerance 
    upper_bound = ideal_angles + tolerance)
    return (angles > lower_bound) & (angles < upper_bound)
```

Then I'd use it here like this:
`(_check_angles(angle1, a['hybridization'], tolerance) & _check_angles(angle2, h['hybridization'], tolerance)).any()`

Also, I'm not sure if this outer-most `any` is actually checking what we want.

Reply:
Okay, I'm pretty sure this `any` should be applied to a single check (separately for `angle1` and `angle2`, so it can be moved to `_check_angles` function), and then we should just combine it with `&`.

Reply:
And then, I think, the current NaNs treatment (i.e. always giving False) will be correct.

Reply:
I like your ideas! Still `any(...)` should de on `-1` axis to account for neighbors for given pair.